examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,308 @@
+  /// the channel name for `C` in the chain it connects to. Handling channels
+  /// should be extracted to a new spec.
+  pure def getCounterparty(moduleName: str, sourceChannel: Channel): Channel = {
+    pure val channels = Map(
+      "A" -> "channelToA",
+      "B" -> "channelToB",
+      "C" -> "channelToC"
+    )
+
+    channels.get(moduleName)
+  }
+
+  /// `sendFugibleTokens` returns the updated module state and the updated set
+  /// of in flight packets. This is the type for its returned value.
+  type SendFungibleTokensResult = {
+    moduleState: ModuleState,
+    inFlightPackets: Set[Packet]
+  }
+
+  pure def sendFungibleTokens(moduleName: str, moduleState: ModuleState, packets: Set[Packet], // this could be an `sendPacket` operator when higher-order operators are supported
+                              denomination: DenomTrace, amount: UINT256,
+                              sender: Address, receiver: Address,
+                              sourcePort: str, sourceChannel: Channel,
+                              timeoutHeight: Height,
+                              timeoutTimestamp: uint64): SendFungibleTokensResult = {
+    val bankResult = if (movingBackAlongTrace({port: sourcePort, channel: sourceChannel}, denomination)) {
+      // burn vouchers
+      (moduleState.bank).BurnCoins(sender, denomination, amount)
+    } else {
+      // escrow tokens
+      val escrowAccount = moduleState.channelEscrowAddresses.get(sourceChannel)
+      moduleState.bank.TransferCoins(sender, escrowAccount, denomination, amount)
+    }
+
+    if (bankResult.success) {
+      val data = { denom: denomination, amount: amount, sender: sender, receiver: receiver, memo: "" }
+
+      // handler.sendPacket
+      val packet = {
+        data: data,
+        sourcePort: sourcePort,
+        sourceChannel: sourceChannel,
+        destPort: "transfer",
+        destChannel: getCounterparty(moduleName, sourceChannel)
+      }
+
+      { moduleState: moduleState.with("bank", bankResult.accounts),
+        inFlightPackets: packets.union(Set(packet)) }
+    } else {
+      { moduleState: moduleState, inFlightPackets: packets }
+    }
+  }
+
+  /// `onRecvPacket` should return an acknowledgment, but it also has to update
+  /// the module state, so the return type is the updated value for the module
+  /// state and the acknowledgment
+  type ReceiveTokenPacketResult = {
+    moduleState: ModuleState,
+    acknowledgement: FungibleTokenPacketAcknowledgement
+  }
+
+  pure def onRecvPacket(moduleState: ModuleState, packet: Packet): ReceiveTokenPacketResult = {
+    val data = packet.data
+    if (movingBackAlongTrace({port: packet.sourcePort, channel: packet.sourceChannel}, data.denom)) {
+      // unescrow tokens to receiver
+      val escrowAccount = moduleState.channelEscrowAddresses.get(packet.destChannel)
+      val receiverDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
```

Comment:
Yeah, this bothers me but I TBH also don't like either of your options.
Returning a pair from an `update` is weird, and might me complicated to
understand, and inlining the logic is something I really want to avoid as that
would be a duplication.

I thought of different option, but I'm not completely sure about it: We could
change `updateTrace` so it receives the boolean value for the result of
`movingBackAlongTrace` instead of calling it. Then we could also remove the
`sourcePort` and `sourceChannel` params of `updateTrace` since those are only
used to call `movingBackAlongTrace`. WDYT?

Reply:
> We could change `updateTrace` so it receives the boolean value for the result of `movingBackAlongTrace` instead of calling it.

I like this option!